search/reverse/lookup with geojson output

mtmail opened the pull request:
Replacement for https://github.com/openstreetmap/Nominatim/pull/707, fixes https://github.com/openstreetmap/Nominatim/issues/118

Follows the JSONv2 output fields (category vs class). For reverse returns a FeatureCollection with one result. Unless no result was found, then it's the same `{"error":"Unable to geocode"}` the other json formats return.

PHP, PHPcs, BDD api tests work. Can't see anything in the error log.

lonvia commented:
There has been https://github.com/geocoders/geocodejson-spec around for a while. I wonder if we should follow that spec for geojson and just add fields where there is additional information. One advantage would be that https://github.com/geocoders/geocoder-tester could then be run against Nominatim.

mtmail commented:
I vote for creating a thin wrapper/proxy instead like those who exist for mapbox, navitia, osmnames already https://github.com/geocoders

My worry is that future versions of the spec change fields (e.g. https://github.com/geocoders/geocodejson-spec/issues/15) and we wouldn't be able to support those without breaking backward compatibility (or supporting a `&version` parameter). There haven't been many changes to the spec but Nominatim will always move slower.

lonvia commented on lib/template/address-json.php:
```diff
@@ -57,4 +57,8 @@
 
-javascript_renderData($aFilteredPlaces);
+if ($sOutputFormat == 'geojson' && !isset($aFilteredPlaces['error'])) {
+    javascript_renderData(jsonResultsToGeoJson([$aFilteredPlaces]));
```
This is a very clever idea but I'm a bit concerned that now the data gets copied around yet another time.

mtmail replied:
I started with having everything in the `search-json.php` template but the various conditions became a mess and a lot harder to test. Later I moved the logic into `Lib.php`

lonvia replied:
What I meant is that I would prefer an extra template instead of transforming the json template yet again.

mtmail commented on lib/lib.php:
```diff
@@ -403,6 +403,32 @@ function getResultDiameter($aResult)
+            'geometry' => [
+                'type' => 'Point',
+                'coordinates' => [(float) $aResult['lon'], (float) $aResult['lat']]
+            ]
```
fixed

mtmail commented on lib/lib.php:
```diff
@@ -403,6 +403,32 @@ function getResultDiameter($aResult)
+                'coordinates' => [(float) $aResult['lon'], (float) $aResult['lat']]
+            ]
+        ];
+        if (isset($aResult['asgeojson'])) {
```
Absolutely. Fixed.

mtmail commented:
Updated:
* `licence` key moved from Features to FeatureCollection
* renamed `boundingbox` to `bbox` and fixed the coordinate ordering (and variable type)
* removed `lat` and `lon` keys

mtmail commented:
Cleanup, squashed commits, forced pushed

mtmail commented:
@lonvia I added a format=geocoderjson as well in separate templates. Not sure how to fill the accuracy for forward geocoding.

lonvia commented:
Looks good but there is a couple of failing BDD tests now.

mtmail commented:
Rebased. BDD tests work.

lonvia commented:
I've merged this now with the tiny change that `addressdetails` is required for `geocodejson` as well to get extended infos like about street, city, etc.  There is still API documentation missing for the new formats.
